### PR TITLE
qa: Add explicit references to related CVE's in p2p_invalid_block test

### PR DESCRIPTION
**Pull request by lucash-dev**
This functional test includes two scenarios that test for regressions of vulnerabilities, but they are only briefly described. There are freely available documents explaining in detail the issues, but without explicit mentions, the developer trying to maintain the code needs an additional step of digging in commit history and PR conversations to figure it out.
Added comments to explicitly mention  CVE-2018-17144 and CVE-2012-2459, for more complete documentation. 
This improves developer experience by making understanding the tests easier.

### CODE REVIEW

**Comment by ken2812221** [comment]
Concept ACK.

**Comment by practicalswift** [comment]
utACK 283235e4b074c29b5fd86db3255e3f70817501f4

Good idea!

**Comment by practicalswift** [comment]
@lucash-dev Would you be able to match other CVE:s against tests and include them in this PR? That would be really useful to have. It would be nice to be able to do `git grep CVE-N` on all issued Bitcoin Core CVE:s to be able to find the corresponding test case and/or fix

Perhaps Bitcoin CVE guru @luke-jr can help in this matching? :-)

**Comment by lucash-dev** [comment]
I’ll do that. Sounds like a good  learning opportunity too.

File: test/functional/p2p_invalid_block.py [hunk 1]
```diff
@@ -53,10 +53,11 @@ def run_test(self):
         block_time = best_block["time"] + 1
 
         # Use merkle-root malleability to generate an invalid block with
-        # same blockheader.
+        # same blockheader (CVE-2012-2459).
         # Manufacture a block with 3 transactions (coinbase, spend of prior
         # coinbase, spend of that spend).  Duplicate the 3rd transaction to
         # leave merkle root and blockheader unchanged but invalidate the block.
+        # For more information on merkle-root malleability see src/consesus/merkle.cpp.
```
Comment (practicalswift) [inline test/functional/p2p_invalid_block.py]: Should be "consensus" :-)

File: test/functional/p2p_invalid_block.py [hunk 2]
```diff
@@ -81,7 +82,7 @@ def run_test(self):
 
         node.p2p.send_blocks_and_test([block2], node, success=False, request_block=False, reject_reason='bad-txns-duplicate')
 
-        # Check transactions for duplicate inputs
+        # Check transactions for duplicate inputs (CVE-2018-17144)
```
Comment (MarcoFalke) [inline test/functional/p2p_invalid_block.py]: This current test case is not complete to check both aspects of the CVE. Would be great if you added the test case for the money printing part of the CVE.

Comment (lucash-dev) [reply test/functional/p2p_invalid_block.py]: I noticed that, when I was studying this source.
TBH I think the test for the other CVE isn’t complete either, as it only checks that the mutated block is rejected, not that the valid block (with same merkle root) is accepted after that. 
I’ll add fixes for both.

Comment (lucash-dev) [reply test/functional/p2p_invalid_block.py]: Added it in a new commit.

**Comment by Empact** [comment]
Concept ACK

**Comment by MarcoFalke** [comment]
Needs rebase (silent conflict)

**Comment by laanwj** [comment]
Concept ACK

**Comment by lucash-dev** [comment]
Coudn't find much info on CVE-2012-3789. @luke-jr could you provide any pointers?

**Comment by lucash-dev** [comment]
Rebased. Also refactored to use the new invalid tx template.

**Comment by fanquake** [comment]
@lucash-dev Could you cleanup the trailing whitespace. That's what's causing Travis to fail:
```
./test/functional/data/invalid_txs.py:204:1: W293 blank line contains whitespace
./test/functional/data/invalid_txs.py:205:1: W293 blank line contains whitespace
./test/functional/data/invalid_txs.py:210:10: E131 continuation line unaligned for hanging indent
```

**Comment by laanwj** [comment]
utACK a342ecd20afc422ece488dca43dd9086dd37798b
Could squash some commits and fix spelling in commit messages

**Comment by lucash-dev** [comment]
> utACK [a342ecd](https://github.com/bitcoin/bitcoin/commit/a342ecd20afc422ece488dca43dd9086dd37798b)
> Could squash some commits and fix spelling in commit messages

Sorry, should've prefixed the PR with "WIP" while I was working adding more comments for more CVEs. It's been a while since I've had time to work on this PR, though. It has also already grown far beyond the original intention (just a couple comments). So probably better if I just fix the commit history, and open a new PR later on when I manage to add comments/tests for the remaining CVEs.
Thanks for reviewing!

**Comment by lucash-dev** [comment]
Fixed the commit history.

File: test/functional/data/invalid_txs.py [hunk 3]
```diff
@@ -174,6 +191,43 @@ def get_tx(self):
...
+    def get_tx(self):
+        tx = CTransaction()
+        vin = self.valid_txin
+        vin.scriptSig = CScript([opcode])
+        tx.vin.append(vin)
+        tx.vout.append(CTxOut(1, basic_p2sh))
+        tx.calc_sha256()
+        return tx
+
+
```
Comment (practicalswift) [inline test/functional/data/invalid_txs.py]: Drop one of the newlines here :-)

**Comment by practicalswift** [comment]
utACK 203c3284eec263ac94339779e36f20a32c77f877 modulo nit and squash

File: test/functional/data/invalid_txs.py (hunk 3 above)
Comment (lucash-dev) [reply test/functional/data/invalid_txs.py]: :man_facepalming:

**Comment by lucash-dev** [comment]
> utACK [203c328](https://github.com/bitcoin/bitcoin/commit/203c3284eec263ac94339779e36f20a32c77f877) modulo nit and squash

Fixed nit

> utACK [203c328](https://github.com/bitcoin/bitcoin/commit/203c3284eec263ac94339779e36f20a32c77f877) modulo nit and squash

Fixed nit. Further squashed to 2 commits. one for adding comments, other for new tests. Let me know if you'd like just one commit. just thought it would make sense to separate these two, but it's all the same to me.
And thanks for reviewing!

**Comment by lucash-dev** [comment]
Rebased.

**Comment by Empact** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/14696/commits/38bfca6bb2ad68719415e9c54a981441052da072

**Comment by laanwj** [comment]
ACK 0c62e3aa73839e97e65a3155e06a98d84b700a1e, checked the CVE numbers, thanks for adding documentation

**Pull request merged**